### ISSUE TITLE
BaseUnityPlugin instances are getting destroyed after awake

### ISSUE THREAD

**Issue by jshepler** [aedenthorn/PlanetCrafterMods#7]
I believe this is what is causing many people to complain that some of your mods aren't working. Update() isn't being called so keypresses are not being detected. In bepinex.cfg, can set HideManagerGameObject=true to address this. I verified this works for a mod I made for myself so would probably work for yours.

You should be able to verify this by adding an OnDestroy() and logging something.

https://github.com/BepInEx/BepInEx/issues/420

**Comment by aedenthorn** [comment]
thanks, I had no idea what was wrong

**Comment by jshepler** [comment]
You bet... I'm hoping there's some code that can be done in Awake() to hide itself (or some other way to address this issue) instead of relying on players knowing they need to modify the cfg file, especially if using a mod manager.

**Comment by aedenthorn** [comment]
I just moved everything to a harmony patch - I think the latest version all work? I was using them without trouble. For BepInEx Config Manager I actually created a new GameObject and did the update stuff there

**Comment by jshepler** [comment]
Might want to do another pass. There's at least one mod that's still using Update() - the delete project.

**Comment by aedenthorn** [comment]
Yeah I didn't look at that one yet; I made it for someone who had a problem a while back

**Comment by akarnokd** [comment]
As part of my mod's install instructions, I asked the players to set `HideManagerGameObject=true`. No need to fight the game's aggressive unloading mechanics on the mod's end.

**Comment by jshepler** [comment]
> As part of my mod's install instructions, I asked the players to set `HideManagerGameObject=true`. No need to fight the game's aggressive unloading mechanics on the mod's end.

That's the thing. It relies on players to read that. Many don't read instructions, especially those already familiar with bepinex mods. Sure, most of those would go back and read the instructions when the mod didn't work and would then set that option.

The other problem is using mod managers, like nexus' vortex. It wouldn't change the bepinex.cfg file when installing a mod, would it?

That's why I thought it would be better to deal with it on the mod's end - so the users wouldn't have to.

**Comment by aedenthorn** [comment]
I wonder if you could use reflection to set it on the fly lol

**Comment by jshepler** [comment]
Maybe, but I like what you did. We can piggy-back a postfix on any existing GameObject's Update(). The BaseUnityPlugin's GameObject might get destroyed, taking the instance's code out of Unity's event cycle, but the **static** harmony patches still run just fine.

**Comment by aedenthorn** [comment]
I can't even find where it checks for HideManagerGameObject in the code

**Comment by akarnokd** [comment]
No idea. Probably would be too late by the time mods run.

I use [this code](https://github.com/akarnokd/ThePlanetCrafterMods/blob/main/LibCommon/BepInExConfigCheck.cs#L21) to check if the `BepInEx.cfg` file has `HideManagerGameObject` set to true.

I put up a warning up instructing the user to do the change.

If you are brave, you could write into this file, setting the flag to true, then force quit the game (restart might not work due to Steam). I personally don't do this as changing global settings without user consent (or knowledge) is a two edged sword.

**Comment by jshepler** [comment]
Little bit of google-fu found https://thunderstore.io/package/bbepis/BepInExPack/

which mentions that the option enables HideAndDontSave which led me to

https://docs.unity3d.com/ScriptReference/Object-hideFlags.html
https://docs.unity3d.com/ScriptReference/HideFlags.HideAndDontSave.html

Maybe try setting `base.hideFlags = HideFlags.HideAndDontSave` in Awake or OnEnable or maybe in the ctor?

**Comment by jshepler** [comment]
> I can't even find where it checks for HideManagerGameObject in the code

I couldn't find it in current master, but did find it in the old 5.4.21 version (which is what we're using) in the Chainloader class (Chainloader.cs). Guess they removed that option for 6.0+

```
internal static readonly ConfigEntry<bool> ConfigHideBepInExGOs = ConfigFile.CoreConfig.Bind(
    "Chainloader", "HideManagerGameObject",

...

ManagerObject = new GameObject("BepInEx_Manager");
if (ConfigHideBepInExGOs.Value)
    ManagerObject.hideFlags = HideFlags.HideAndDontSave;

```

**Comment by jshepler** [comment]
Doing `this.gameObject.hideFlags = HideFlags.HideAndDontSave;` in Awake or OnEnable works.

BTW, they did remove the option in the current version of bepinex, but they always set the flag: https://github.com/BepInEx/BepInEx/blob/master/Runtimes/Unity/BepInEx.Unity.Mono/Bootstrap/UnityChainloader.cs#L100
